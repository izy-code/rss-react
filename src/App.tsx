import type { ReactNode } from 'react';
import { Component } from 'react';

import { MainPage } from '@/pages/main-page/MainPage';

export class App extends Component {
  public render(): ReactNode {
    return <MainPage />;
  }
}
