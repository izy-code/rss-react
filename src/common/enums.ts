export const enum RoutePath {
  MAIN = '/',
  UNCONTROLLED_FORM = '/uncontrolled-form',
  REACT_HOOK_FORM = '/react-hook-form',
}
