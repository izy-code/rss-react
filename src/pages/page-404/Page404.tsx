import { useNavigate } from '@remix-run/react';
import type { ReactNode } from 'react';

import { CustomButton } from '@/components/custom-button/CustomButton';

import styles from './Page404.module.scss';

export function Page404(): ReactNode {
  const navigate = useNavigate();

  return (
    <main className={styles.main}>
      <div className={styles.container}>
        <h1 className={styles.title}>404</h1>
        <p className={styles.text}>The page you requested was not found.</p>
        <div className={styles.buttons}>
          <CustomButton variant="tertiary" onClick={() => navigate(-1)}>
            Previous page
          </CustomButton>
          <CustomButton
            variant="tertiary"
            onClick={() => {
              const hostUrl = `${window.location.protocol}//${window.location.host}`;

              window.location.href = hostUrl;
            }}
          >
            Last stored search
          </CustomButton>
        </div>
      </div>
    </main>
  );
}
