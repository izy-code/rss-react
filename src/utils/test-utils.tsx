import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';
import type { NextRouter } from 'next/router';
import { type FC, type PropsWithChildren, type ReactElement, type ReactNode, useRef } from 'react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/store';
import { makeStore, setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProvidersAndUser(
  ui: ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState), ...renderOptions }: ExtendedRenderOptions = {},
): { store: AppStore; user: ReturnType<typeof userEvent.setup>; container: HTMLElement } {
  function Wrapper({ children }: PropsWithChildren): ReactElement {
    const storeRef = useRef<AppStore>();

    if (!storeRef.current) {
      storeRef.current = makeStore();
    }

    return <Provider store={storeRef.current}>{children}</Provider>;
  }

  return { store, user: userEvent.setup(), ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithUserSetup(ui: ReactElement): {
  user: ReturnType<typeof userEvent.setup>;
  container: HTMLElement;
} {
  return {
    user: userEvent.setup(),
    ...render(ui),
  };
}

function createMockedNextRouter(router: Partial<NextRouter>): NextRouter {
  return {
    basePath: '',
    pathname: '/',
    route: '/',
    query: {},
    asPath: '/',
    back: vi.fn(),
    beforePopState: vi.fn(),
    prefetch: vi.fn(),
    push: vi.fn(),
    reload: vi.fn(),
    replace: vi.fn(),
    events: {
      on: vi.fn(),
      off: vi.fn(),
      emit: vi.fn(),
    },
    isFallback: false,
    isLocaleDomain: false,
    isReady: true,
    defaultLocale: 'en',
    domainLocales: [],
    isPreview: false,
    forward: vi.fn(),
    ...router,
  };
}

export const getMockedNextRouter: (nextRouterProps?: Partial<NextRouter>) => {
  nextRouter: NextRouter;
  NextProvider: FC<{ children: ReactNode }>;
} = (nextRouterProps = {}) => {
  const nextRouter = createMockedNextRouter(nextRouterProps);

  return {
    nextRouter,
    NextProvider: ({ children }) => <RouterContext.Provider value={nextRouter}>{children}</RouterContext.Provider>,
  };
};
