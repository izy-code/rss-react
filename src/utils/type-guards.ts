import type { CharacterData, CharacterListData, CharacterListInfo } from '@/api/types';

export function isCharacterData(data: unknown): data is CharacterData {
  return Boolean(
    typeof data === 'object' &&
      data &&
      'id' in data &&
      typeof data.id === 'number' &&
      'name' in data &&
      typeof data.name === 'string' &&
      'image' in data &&
      typeof data.image === 'string' &&
      'species' in data &&
      typeof data.species === 'string' &&
      'status' in data &&
      typeof data.status === 'string' &&
      'gender' in data &&
      typeof data.gender === 'string' &&
      'url' in data &&
      typeof data.url === 'string' &&
      'created' in data &&
      typeof data.created === 'string' &&
      'episode' in data &&
      Array.isArray(data.episode) &&
      data.episode.every((item: unknown) => typeof item === 'string') &&
      'origin' in data &&
      typeof data.origin === 'object' &&
      data.origin &&
      'name' in data.origin &&
      typeof data.origin.name === 'string' &&
      'url' in data.origin &&
      typeof data.origin.url === 'string' &&
      'location' in data &&
      typeof data.location === 'object' &&
      data.location &&
      'name' in data.location &&
      typeof data.location.name === 'string' &&
      'url' in data.location &&
      typeof data.location.url === 'string' &&
      'type' in data &&
      typeof data.type === 'string',
  );
}

function isCharacterListInfoData(data: unknown): data is CharacterListInfo {
  return Boolean(
    typeof data === 'object' &&
      data &&
      'count' in data &&
      typeof data.count === 'number' &&
      'pages' in data &&
      typeof data.pages === 'number' &&
      'next' in data &&
      (typeof data.next === 'string' || data.next === null) &&
      'prev' in data &&
      (typeof data.prev === 'string' || data.prev === null),
  );
}

export function isCharacterListData(data: unknown): data is CharacterListData {
  return Boolean(
    typeof data === 'object' &&
      data &&
      'info' in data &&
      isCharacterListInfoData(data.info) &&
      'results' in data &&
      Array.isArray(data.results) &&
      data.results.every((item: unknown) => isCharacterData(item)),
  );
}
