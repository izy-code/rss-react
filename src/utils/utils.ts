export type DataObject = Record<string, string | Record<string, string>>;

export const getCsvObjectUrl = (data: DataObject[]): string => {
  if (data.length === 0 || !data[0]) {
    return '';
  }

  const flattenObject = (
    obj: DataObject,
    parentKey = '',
    result: Record<string, string> = {},
  ): Record<string, string> => {
    Object.entries(obj).forEach(([key, value]) => {
      const newKey = parentKey ? `${parentKey}.${key}` : key;
      if (key === 'episode' && typeof value === 'object') {
        const episodeValues = Object.values(value);
        result[newKey] = episodeValues.join(', ');
      } else if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        flattenObject(value, newKey, result);
      } else {
        result[newKey] = String(value);
      }
    });

    return result;
  };

  const titlesArray = Object.keys(flattenObject(data[0]));
  const rowsArray = [titlesArray];

  data.forEach((item) => {
    const flattenedItem = flattenObject(item);
    rowsArray.push(Object.values(flattenedItem));
  });

  let csvString = '';

  rowsArray.forEach((row) => {
    csvString += `${row.join(';')}\n`;
  });

  const blob = new Blob([csvString], { type: 'text/csv;charset=utf-8' });

  return URL.createObjectURL(blob);
};
