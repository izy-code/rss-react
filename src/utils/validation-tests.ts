import type { TestFunction } from 'yup';

import { ALLOWED_FILE_TYPES, IMAGE_MAX_SIZE_BYTES } from '@/common/constants';

export const isAllowedFileSize: TestFunction<unknown> = (file: unknown) => {
  if (!(file instanceof File)) {
    return false;
  }

  return file.size <= IMAGE_MAX_SIZE_BYTES;
};

export const isAllowedFileType: TestFunction<unknown> = (file: unknown) => {
  if (!(file instanceof File)) {
    return false;
  }

  return ALLOWED_FILE_TYPES.includes(file.type);
};
