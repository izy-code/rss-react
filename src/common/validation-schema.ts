import type { InferType } from 'yup';
import { boolean, mixed, number, object, ref, string } from 'yup';

import { isAllowedFileSize, isAllowedFileType } from '@/utils/validation-tests';

import {
  COUNTRIES,
  LOWERCASE_LETTER_REGEX,
  MAX_AGE,
  NUMBER_REGEX,
  SPECIAL_CHARACTER_REGEX,
  UPPERCASE_LETTER_REGEX,
} from './constants';

export const validationSchema = object().shape({
  name: string()
    .required('Name is required')
    .matches(/^\p{Lu}/u, 'Name must start with a capital letter'),
  age: number()
    .transform((parsedValue, originalValue) => (originalValue === '' ? null : parsedValue) as number | null)
    .required('Age is required')
    .min(0, `Age must be non negative`)
    .max(MAX_AGE, `Age must be less than ${MAX_AGE}`)
    .integer(`Age must be an integer`),
  email: string().required('E-mail is required').email('Email must have valid format'),
  password: string()
    .required('Password is required')
    .matches(NUMBER_REGEX, 'Password must contain a number')
    .matches(UPPERCASE_LETTER_REGEX, 'Password must contain an uppercase letter')
    .matches(LOWERCASE_LETTER_REGEX, 'Password must contain a lowercase letter')
    .matches(SPECIAL_CHARACTER_REGEX, 'Password must contain a character from "-+/%*:#@\\$!?|^&"'),
  'password-confirm': string()
    .required('Please confirm your password')
    .oneOf([ref('password')], 'Passwords do not match'),
  gender: string().oneOf(['male', 'female', 'other']).required(),
  tac: boolean()
    .required('You must accept the Terms and Conditions')
    .isTrue('You must accept the Terms and Conditions'),
  picture: mixed((input): input is File => input instanceof File)
    .transform((value) => {
      if (value instanceof File && value.size > 0) {
        return value;
      }
      if (value instanceof FileList && value.item(0) && value.item(0)!.size > 0) {
        return value.item(0);
      }
      return undefined;
    })
    .required('Image is required')
    .defined('Image is required')
    .test('file-size', 'Image size should be less than 2 MB', isAllowedFileSize)
    .test('file-type', 'Allowed image types: PNG, JPG', isAllowedFileType),
  country: string()
    .transform((parsedValue, originalValue) => (originalValue === '' ? null : parsedValue) as string | null)
    .required('Country is required')
    .oneOf(COUNTRIES, 'Should be a country from the list'),
});

export type SchemaInferredType = InferType<typeof validationSchema>;
