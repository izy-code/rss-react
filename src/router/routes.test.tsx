import type { TypedResponse } from '@remix-run/node';
import { json } from '@remix-run/node';
import { createRemixStub } from '@remix-run/testing';
import { screen } from '@testing-library/react';
import type { ReactNode } from 'react';

import type { FetchCharacterListResult, FetchCharacterResult } from '@/api/api';
import App from '@/app/root';
import Index from '@/app/routes/_index';
import { SearchParams } from '@/common/enums';
import { Page404 } from '@/pages/page-404/Page404';
import { characterMock, charactersDataMock } from '@/test/mocks/mocks';
import { renderWithProvidersAndUser } from '@/utils/test-utils';

const AppRemixStub = createRemixStub([
  {
    id: 'root',
    path: '/',
    Component: (): ReactNode => <App />,
    children: [
      {
        id: 'index',
        path: '/',
        Component: (): ReactNode => <Index />,
        loader: (): TypedResponse<FetchCharacterResult> => json({ status: 'success', data: characterMock }),
      },
    ],
    loader: (): TypedResponse<FetchCharacterListResult> => json({ status: 'success', data: charactersDataMock }),
  },
]);

describe('Router render', () => {
  it('should match the snapshot for main route', async () => {
    const { container } = renderWithProvidersAndUser(
      <AppRemixStub
        hydrationData={{
          loaderData: {
            root: { status: 'success', data: charactersDataMock },
            index: { status: 'success', data: characterMock },
          },
        }}
        initialEntries={[`/?${SearchParams.DETAILS}=${characterMock.id}`]}
      />,
    );

    await screen.findByRole('button', { name: /prev/i });

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot for 404 page', () => {
    const RemixStub404 = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => <Page404 />,
      },
    ]);

    const { container } = renderWithProvidersAndUser(<RemixStub404 />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot for error page', async () => {
    const consoleErrorMock = vi.spyOn(console, 'error').mockImplementation(() => {});

    const { container, user } = renderWithProvidersAndUser(
      <AppRemixStub
        hydrationData={{
          loaderData: {
            root: { status: 'success', data: charactersDataMock },
            index: { status: 'success', data: characterMock },
          },
        }}
        initialEntries={[`/?${SearchParams.DETAILS}=${characterMock.id}`]}
      />,
    );

    const errorButton = screen.getByRole('button', { name: /throw error/i });

    await user.click(errorButton);

    expect(consoleErrorMock).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
