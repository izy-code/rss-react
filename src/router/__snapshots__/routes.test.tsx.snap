// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Router render > should match the snapshot for 404 page 1`] = `
<div>
  <main
    class="_main_1e53a6"
  >
    <div
      class="_container_1e53a6"
    >
      <h1
        class="_title_1e53a6"
      >
        404
      </h1>
      <p
        class="_text_1e53a6"
      >
        The page you requested was not found.
      </p>
      <div
        class="_buttons_1e53a6"
      >
        <button
          class="_button_1c0f93 _button--tertiary_1c0f93"
          type="button"
        >
          Previous page
        </button>
        <button
          class="_button_1c0f93 _button--tertiary_1c0f93"
          type="button"
        >
          Home page
        </button>
      </div>
    </div>
  </main>
</div>
`;

exports[`Router render > should match the snapshot for error page 1`] = `
<div>
  <main
    class="_main_6e3859"
  >
    <div
      class="_container_6e3859"
    >
      <h1
        class="_header_6e3859"
      >
        Oops!
      </h1>
      <p
        class="_text_6e3859"
      >
        Sorry, an unexpected error has occurred.
      </p>
      <p
        class="_errorDesc_6e3859"
      >
        Error message:
      </p>
      <p
        class="_error_6e3859"
      >
        Error throwing button was clicked
      </p>
      <p
        class="_text_6e3859"
      >
        Please try to refresh the page.
      </p>
      <button
        class="_button_1c0f93 _button--primary_1c0f93 _refreshBtn_6e3859"
        type="button"
      >
        Refresh the page
      </button>
    </div>
  </main>
</div>
`;

exports[`Router render > should match the snapshot for main routes 1`] = `
<div>
  <div
    class="_page_6798d4"
  >
    <header
      class="_header_accb48"
    >
      <div
        class="_container_accb48"
      >
        <h1
          class="visually-hidden"
        >
          Rick and Morty characters
        </h1>
        <form
          class="_form_1de04e"
        >
          <input
            autocomplete="off"
            class="_input_1de04e"
            placeholder="Enter character nameâ€¦"
            type="search"
            value=""
          />
          <button
            class="_button_1c0f93 _button--secondary_1c0f93"
            type="submit"
          >
            Search
          </button>
        </form>
        <button
          class="_button_1c0f93 _button--cancel_1c0f93"
          type="button"
        >
          Throw error
        </button>
      </div>
    </header>
    <main
      class="_main_6798d4"
    >
      <section
        class="_listSection_6798d4"
      >
        <div
          class="_loaderContainer_523128"
        >
          <h2
            class="visually-hidden"
          >
            Loading...
          </h2>
          <div
            class="_loader_523128"
          />
        </div>
      </section>
    </main>
  </div>
</div>
`;
